### PR TITLE
Run fi_rdm_tagged_bw Test with Fork

### PR DESCRIPTION
**Pull request by aolia-0303**
Currently in fabtests, when running fi_rdm_tagged_bw, we can specify
a flag `-K` to let the sender and receiver fork a child process after
initializing endpoint. This commit adds logic that runs
fi_rdm_tagged_bw with `-K` flag with different environment
variables set in multinode_runfabtests.sh

Signed-off-by: Ao Li <aolia@amazon.com>

### CODE REVIEW

File: multinode_runfabtests.sh [hunk 1]
```diff
@@ -37,6 +37,63 @@ run_dgram_pingpong_with_b()
...
+        rdm_tagged_bw_cmd="${HOME}/libfabric/fabtests/install/bin/fi_rdm_tagged_bw -p efa -K -E"
+    elif [ ${FORK_MODE} = "RDMAV_FORK_SAFE" ]; then
+        FORK_ENV="RDMAV_FORK_SAFE=1"
+        rdm_tagged_bw_cmd="${HOME}/libfabric/fabtests/install/bin/fi_rdm_tagged_bw -v -p efa -K -E"
+    elif [ ${FORK_MODE} = "FI_EFA_FORK_SAFE" ]; then
+        FORK_ENV="FI_EFA_FORK_SAFE=1"
+        rdm_tagged_bw_cmd="${HOME}/libfabric/fabtests/install/bin/fi_rdm_tagged_bw -v -p efa -K -E"
+    fi
+
+    ssh -o ConnectTimeout=30 -o StrictHostKeyChecking=no ${SERVER_IP} ${FORK_ENV} ${rdm_tagged_bw_cmd} >& server.out &
```
Comment (wzamazon) [inline multinode_runfabtests.sh:57]: Do ${FORK_ENV} actually set the environment variable on remote instance?

File: multinode_runfabtests.sh [hunk 2]
```diff
@@ -37,6 +37,63 @@ run_dgram_pingpong_with_b()
     return $ret
 }
 
+run_rdm_tagged_bw_fork()
```
Comment (wzamazon) [inline multinode_runfabtests.sh:40]: This function is a duplicate of `run_dgram_pingpong_with_b`, can we come up we with a common function for both?

File: multinode_runfabtests.sh [hunk 3]
```diff
@@ -77,4 +134,16 @@ if [ ${PROVIDER} == "efa" ]; then
     # So here we run it separately using "-b" option
     echo "Run fi_dgram_pingpong with out-of-band synchronization"
     run_dgram_pingpong_with_b ${SERVER_IP} ${CLIENT_IP}
+
+    # Run fi_rdm_tagged_bw with fork when different environment variables are set.
+
+    echo "Run fi_rdm_tagged_bw with fork"
+    run_rdm_tagged_bw_fork ${SERVER_IP} ${CLIENT_IP}
```
Comment (wzamazon) [inline multinode_runfabtests.sh:141]: With out setting environment variables, this test will actually fail, which is expected. How to account for that?

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: multinode_runfabtests.sh (hunk 3 above)
Comment (aolia-0303) [reply multinode_runfabtests.sh:141]: We expect non zero return status when no environment variables is set "https://github.com/awslabs/libfabric-ci-scripts/pull/140/files#diff-effebc9b96458b408492a90d2789d72f1e79cc496af4392c51cc52e595686e6dR75".

File: multinode_runfabtests.sh (hunk 1 above)
Comment (aolia-0303) [reply multinode_runfabtests.sh:57]: `FORK_ENV` sets the environment variable only during the duration when `fi_rdm_tagged_bw` is run. 

For example, if we call `run_rdm_tagged_bw_fork ${SERVER_IP} ${CLIENT_IP}` after `run_rdm_tagged_bw_fork ${SERVER_IP} ${CLIENT_IP} "FI_EFA_FORK_SAFE"`. No environment variables will be set.

Comment (wzamazon) [reply multinode_runfabtests.sh:57]: You are using ssh to execute the command remotely, right? and you want to set the environment variable on remote machine, not local machine. My question is did the command line actually set the environment variable on the remote machine? My experience has been that command line like the following:

`ssh user@host ENV=xxx programs` will not set `ENV` to `xxx` on the remote machine.

File: multinode_runfabtests.sh (hunk 2 above)
Comment (aolia-0303) [reply multinode_runfabtests.sh:40]: I think we can. But my concern is that we might need many extra parameters, like the command that we would like to run, whether the test is expected to pass or fail, etc.

File: multinode_runfabtests.sh (hunk 1 above)
Comment (aolia-0303) [reply multinode_runfabtests.sh:57]: I do not think there is a need to set environment variable on remote machine. By using the syntax `ssh user@host ENV=xxx programs`, we only temporary sets the `ENV` to `xxx` when running `programs`.

File: multinode_runfabtests.sh (hunk 2 above)
Comment (wzamazon) [reply multinode_runfabtests.sh:40]: That is still preferred than duplicating a whole bunch of code.

File: multinode_runfabtests.sh (hunk 1 above)
Comment (wzamazon) [reply multinode_runfabtests.sh:57]: But the application `fi_rdm_tagged_bw` is run on remote machine, and its behavior would be different depend on the environment variable, right? which is what we are testing?

File: multinode_runfabtests.sh (hunk 2 above)
Comment (aolia-0303) [reply multinode_runfabtests.sh:40]: Ok. And I will add comments regarding how to call the new function as well.

File: multinode_runfabtests.sh (hunk 1 above)
Comment (wzamazon) [reply multinode_runfabtests.sh:57]: Never mind. I verified that ssh will pass env to remote machine.

File: multinode_runfabtests.sh (hunk 2 above)
Comment (wzamazon) [reply multinode_runfabtests.sh:40]: Now that we have `run_test_with_expected_ret`. Do we still need this function?

Comment (wzamazon) [reply multinode_runfabtests.sh:40]: Also the function `run_dgram_pingpong_with_b` can be removed. Right?

File: multinode_runfabtests.sh [hunk 4]
```diff
@@ -69,12 +175,23 @@ if [ ${PROVIDER} == "efa" ]; then
...
     # dgram_pingpong test has been excluded during installation
     # (in install-fabtests.sh), because it does not work with "-E" option.
     # So here we run it separately using "-b" option
     echo "Run fi_dgram_pingpong with out-of-band synchronization"
-    run_dgram_pingpong_with_b ${SERVER_IP} ${CLIENT_IP}
+    run_test_with_expected_ret ${SERVER_IP} ${CLIENT_IP} "${HOME}/libfabric/fabtests/install/bin/fi_dgram_pingpong" "PASS"
+
+    # Run fi_rdm_tagged_bw with fork when different environment variables are set.
+    echo "Run fi_rdm_tagged_bw with fork"
+    run_test_with_expected_ret ${SERVER_IP} ${CLIENT_IP} "${HOME}/libfabric/fabtests/install/bin/fi_rdm_tagged_bw -p efa -K -E" "FAIL"
```
Comment (wzamazon) [inline multinode_runfabtests.sh:190]: My understanding is this script will be run by another script with `set -e`. So when one process failed, the whole test will immediately fail. Can you verify if it is the case?

Comment (aolia-0303) [reply multinode_runfabtests.sh:190]: Yeah. In the case of `set -e`. If the one process failed, the whole test will immediately fail.

Comment (wzamazon) [reply multinode_runfabtests.sh:190]: Please refer how other scripts handle this issue. For example, `wget_check.sh` will `set +e` if `set -e` has been called, and restore the option e at the end if necessary.

Comment (aolia-0303) [reply multinode_runfabtests.sh:190]: Sure.

**Review by wzamazon — APPROVED** [review]

**Pull request merged**